use_cluster.py

```diff
@@ -366,4 +355,391 @@ def return_cluster_dict(q_cluster, q_ids_list):
+    global_keywords = list(set([x.upper() for x in global_keywords]))
+    return global_keywords
+
+
+
+"""
+> The objective of this function is to:
+1. Calculate the distances for each keyword from the centroid, to be used for scoring
+    
+2. Call the custom metric function to calculate scores for labels 
+    > Call NMI_Metric
+    > Add distances term with the factor to normalise, to get the custome score
+
+3. Return list of Top 5 Label Scores for each cluster
+
+:The inputs required are taken from the outputs of the Generate_Modified_Qs_Data function 
+: The cluster object has attributes: centroid of cluster, raw text questions, lemmatized questions
+"""
+
+def Calculate_Label_Score(cluster_objects, vecs, global_keywords, lemmatized_qs_list):
+    #For each term and each cluster, get frequencies over all qs present in our corpus (NMI)
+    num_clus = len(cluster_objects)
+    all_cluster_scores = []
+    for clus in range(num_clus):
+        NMI = [] #NMI Scores from custom metric (with Centroid Factor)
+        vanilla_NMI=[] #Vanilla NMI Scores
+        modified_clus_qs = cluster_objects[clus]["keyword_questions"] #clus_qs list of qs from the modified cluster 
+        og_clus_qs = cluster_objects[clus]["original clus qs"]
+        centroid = cluster_objects[clus]["centroid"]
+        vec_to_centroid = centroid - vecs
+        distances = np.linalg.norm(vec_to_centroid, axis=1)
+        logging.debug("Total number of Keywords:",len(distances))
+        gloabl_min_distance = min(distances)
+        logging.debug("Global Minimum Distance for cluster:",gloabl_min_distance)
+    
+        NMI_Metric(global_keywords, lemmatized_qs_list, modified_clus_qs, NMI, vanilla_NMI)
+
+        vanilla_NMI.sort(key = lambda x: x[1], reverse=True)
+        max_NMI = max(NMI,key=lambda x:x[1])[1]
+        logging.debug("Maximum NMI Score in Cluster:",max_NMI)
+        logging.debug("Length of score list:",len(NMI))
+        assert_equal(len(NMI),len(distances)) #Making sure equal lenghts for math to work out xD
+        new_score = [[x[0],x[1] + (1/d)*gloabl_min_distance*max_NMI*0.5] for x,d in zip(NMI,distances)]
+        new_score.sort(key = lambda x: x[1], reverse=True)
+        logging.debug("Old NMI Score based Labels:",vanilla_NMI[:5])
+        logging.debug("New Score based Labels:",new_score[:5])
+        logging.debug("Cluster Qs:",og_clus_qs)
+        all_cluster_scores.append(new_score[:5])
+    return all_cluster_scores
+
+
+"""
+> The objective of this function is to:
+1. Calculate the Normalised Mutual Information (NMI) score for a term & a cluster
+    
+2. Append all scores in both NMI & Vanilla NMI Lists for final calculation & comparision
+        
+:The inputs required are the global keywords list, lemmatized question list, list of cluster questions
+and the NMI & Vanilla NMI lists, which are modified at each call of this function  
+"""
+
+
+def NMI_Metric(global_keywords, lemmatized_qs_list, clus_qs, NMI, vanilla_NMI):
```

Comment:
Hey Jeff so I had seen this, and I had a similar doubt. So in the paper they didn't mention how and where are they accounting for this negative sign. I implemented the algorithm myself, and then tested to see if it gives me any word with a high score which is not in the cluster, and I couldn't find in any of the examples we used (this was without the centroid metric). So maybe that is not happening, or the implementation I did is taking care of that, or maybe we didn't catch it. 

But I agree, I wanted more clarity on this, and I would want to explore on how they are identifying those associations and assigning it a negative sign. 

One way to make sure we never select a label which is not part of the cluster, is to have a condition over the final list which removes any label from top 10 (coz we are only considering top 10 at the end for each cluster) which isn't actually present in the cluster questions.

I would love to discuss more on this with you :D

Reply:
I think your "remove any label not in the cluster" solution is good! And it's probably a minor thing that wouldn't come up much given you couldn't find any examples. But I think it's still worth putting in that solution as a safeguard.

Reply:
Also, I would include a comment above the portion of the code that removes labels not in the cluster explaining why we need to do it - otherwise, I think it could be confusing to somebody who hasn't read the paper.

Reply:
So as I updated you on slack, have figured out the conditions from the paper, and have implemented them! :D 
Have also added comments where I added this new logic 👍